perlPackages.ClusterSSH: init at 4.13.2

aither64 opened the pull request:
###### Motivation for this change
Adds ClusterSSH with one dependency and myself as a maintainer. ClusterSSH is a tool that lets you interactively manage multiple servers over SSH at once.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

aither64 committed:
perlPackages.X11ProtocolOther: init at 30

aither64 committed:
perlPackages.ClusterSSH: init at 4.13.2

aither64 committed:
maintainers: add aither

volth commented:
May be from CPAN instead of GitHub to simplify auto-update ?
It has the same version https://metacpan.org/pod/App::ClusterSSH::Cluster

aither64 commented:
You're right, thanks! I've been looking at CPAN before, but I thought it's just some library and not the program itself.

I've also added installation of script for bash completion and missing dependency on openssh.

aanderse commented:
@GrahamcOfBorg build perlPackages.ClusterSSH

@aither64 Doing a quick scan of the perl source code I found this: https://metacpan.org/source/DUNCS/App-ClusterSSH-v4.13.2/lib/App/ClusterSSH/Config.pm#L464

Without looking into it in detail, did you test the application in such a way that you confirmed this code works as intended?

GrahamcOfBorg commented:
<!--REQUEST_ID=b2995a12-1a10-4730-9f5b-0f6bd406e04b-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=6590eee6-7f1e-45a1-9d66-030ddc4958ee)

Attempted: perlPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/wbp00fgb02vyq50gavpa7xa5amch0q2h-perl5.28.0-App-ClusterSSH-4.13.2/lib  /nix/store/wbp00fgb02vyq50gavpa7xa5amch0q2h-perl5.28.0-App-ClusterSSH-4.13.2/bin
patching script interpreter paths in /nix/store/wbp00fgb02vyq50gavpa7xa5amch0q2h-perl5.28.0-App-ClusterSSH-4.13.2
checking for references to /build in /nix/store/wbp00fgb02vyq50gavpa7xa5amch0q2h-perl5.28.0-App-ClusterSSH-4.13.2...
shrinking RPATHs of ELF executables and libraries in /nix/store/2zsrgyrs6xiikz1191rs6vbnk5pr3pbk-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
gzipping man pages under /nix/store/2zsrgyrs6xiikz1191rs6vbnk5pr3pbk-perl5.28.0-App-ClusterSSH-4.13.2-devdoc/share/man/
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/2zsrgyrs6xiikz1191rs6vbnk5pr3pbk-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
checking for references to /build in /nix/store/2zsrgyrs6xiikz1191rs6vbnk5pr3pbk-perl5.28.0-App-ClusterSSH-4.13.2-devdoc...
/nix/store/wbp00fgb02vyq50gavpa7xa5amch0q2h-perl5.28.0-App-ClusterSSH-4.13.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=b2995a12-1a10-4730-9f5b-0f6bd406e04b-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=a063b422-c968-47d5-9aaa-e168001aaf45)

Attempted: perlPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/lqsvcdn8lpwzdivx458f4dbambqwxvhl-perl5.28.0-App-ClusterSSH-4.13.2/lib  /nix/store/lqsvcdn8lpwzdivx458f4dbambqwxvhl-perl5.28.0-App-ClusterSSH-4.13.2/bin
patching script interpreter paths in /nix/store/lqsvcdn8lpwzdivx458f4dbambqwxvhl-perl5.28.0-App-ClusterSSH-4.13.2
checking for references to /build in /nix/store/lqsvcdn8lpwzdivx458f4dbambqwxvhl-perl5.28.0-App-ClusterSSH-4.13.2...
shrinking RPATHs of ELF executables and libraries in /nix/store/l0haqanl5825gypqfrdv5641xp6naid8-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
gzipping man pages under /nix/store/l0haqanl5825gypqfrdv5641xp6naid8-perl5.28.0-App-ClusterSSH-4.13.2-devdoc/share/man/
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/l0haqanl5825gypqfrdv5641xp6naid8-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
checking for references to /build in /nix/store/l0haqanl5825gypqfrdv5641xp6naid8-perl5.28.0-App-ClusterSSH-4.13.2-devdoc...
/nix/store/lqsvcdn8lpwzdivx458f4dbambqwxvhl-perl5.28.0-App-ClusterSSH-4.13.2
```
</p></details>

Mic92 commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -2327,6 +2327,37 @@ let
+    preConfigure = ''
+      for fn in bin_PL/* ; do
+        substituteInPlace "$fn" --replace '/usr/bin/env perl' '/usr/bin/perl'
+      done
```
the `patchShebangs` function should fix this also

Mic92 replied:
Ok. Wait a sec, why are you replacing it to `/usr/bin/perl`?

aither64 replied:
This is what happens without it:
```
$ cssh
/nix/store/dzvbhkzfgwkq97ljkcvi4gs7vmflh6a0-coreutils-8.30/bin/env: ‘perl -I/nix/store/0ingn8cwwnl84i374hcl6nafsm2c5m2p-perl-5.’: No such file or directory
/nix/store/dzvbhkzfgwkq97ljkcvi4gs7vmflh6a0-coreutils-8.30/bin/env: use -[v]S to pass options in shebang lines
```
As I understand it, Linux allows only one argument to shebang, so `env` takes `perl` and all its path arguments as a single argument and tries to execute it. Patching to `/usr/bin/perl` solves this, I guess perl handles this situation. I can change it if you have a better idea.

Mic92 commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -2327,6 +2327,37 @@ let
+      sha256 = "0rmk2p3f2wz1h092anidjclh212rv3gxyk0c641qk3frlrjnw6mp";
+    };
+    propagatedBuildInputs = [ ExceptionClass FilePath GetoptLong LocaleMaketext Tk TryTiny X11Protocol X11ProtocolOther pkgs.coreutils pkgs.openssh pkgs.xterm ];
+    buildInputs = [ FileSlurp FileTemp FileWhich CPANChanges TestPerlTidy TestPod TestPodCoverage Readonly TestDifferences TestDistManifest TestTrap ];
```
Please break both lists into multiple lines for readability.

aither64 replied:
Done

volth replied:
@Mic92: these lists are managed by ```cpan2nix``` so it will sort them and put into single lines on the next run (and would undo any manual changes which are not in the exception list).
I see the problem with readability and probably ```cpan2nix``` should address it, but perhaps it would be better to tag somehow the parts of file as bot-generated.

Mic92 replied:
ok. I was not aware of that.

aither64 commented:
@aanderse, yes, I'm using it and it works for me. The code you linked is used to find xterm (or some other configured terminal), so the hardcoded paths would be used if it weren't found already in `$PATH`, i.e. it will not find it on NixOS and exit. I guess we could add `/run/current-system/sw/bin` to the hardcoded list, but I think it is sufficient to rely on `$PATH` being set properly and having the terminal in it.

GrahamcOfBorg commented:
<!--REQUEST_ID=fd006f54-2094-4919-b808-1e703f2707b4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=6322dd7a-95fa-45ee-a3dd-4dfeb7eb12f4)

Attempted: perlPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/xsn0s5g49650y89241vdal28wz1l7knk-perl5.28.0-App-ClusterSSH-4.13.2/lib  /nix/store/xsn0s5g49650y89241vdal28wz1l7knk-perl5.28.0-App-ClusterSSH-4.13.2/bin
patching script interpreter paths in /nix/store/xsn0s5g49650y89241vdal28wz1l7knk-perl5.28.0-App-ClusterSSH-4.13.2
checking for references to /build in /nix/store/xsn0s5g49650y89241vdal28wz1l7knk-perl5.28.0-App-ClusterSSH-4.13.2...
shrinking RPATHs of ELF executables and libraries in /nix/store/gzynavpxaw0d87w3qk0yg23ddk26h39b-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
gzipping man pages under /nix/store/gzynavpxaw0d87w3qk0yg23ddk26h39b-perl5.28.0-App-ClusterSSH-4.13.2-devdoc/share/man/
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/gzynavpxaw0d87w3qk0yg23ddk26h39b-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
checking for references to /build in /nix/store/gzynavpxaw0d87w3qk0yg23ddk26h39b-perl5.28.0-App-ClusterSSH-4.13.2-devdoc...
/nix/store/xsn0s5g49650y89241vdal28wz1l7knk-perl5.28.0-App-ClusterSSH-4.13.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=fd006f54-2094-4919-b808-1e703f2707b4-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=2ac945f0-c58c-4fe3-873f-7c89b45082b4)

Attempted: perlPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
  TODO passed:   160-163, 168-171
t/30cluster.t   (Wstat: 512 Tests: 59 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
Files=16, Tests=1916, 11 wallclock secs ( 0.23 usr  0.09 sys +  8.08 cusr  2.26 csys = 10.66 CPU)
Result: FAIL
Failed 1/16 test programs. 0/1916 subtests failed.
make: *** [Makefile:1025: test_dynamic] Error 255
builder for '/nix/store/csp92ks9w95rm3p9pkl60w9qsdq9bc33-perl5.28.0-App-ClusterSSH-4.13.2.drv' failed with exit code 2
error: build of '/nix/store/csp92ks9w95rm3p9pkl60w9qsdq9bc33-perl5.28.0-App-ClusterSSH-4.13.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=fd006f54-2094-4919-b808-1e703f2707b4-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=fc8215dd-de6b-44d2-a647-ef4dc8474c04)

Attempted: perlPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/58lissxzppdn2cy62abhfq98ln83lvqd-perl5.28.0-App-ClusterSSH-4.13.2/lib  /nix/store/58lissxzppdn2cy62abhfq98ln83lvqd-perl5.28.0-App-ClusterSSH-4.13.2/bin
patching script interpreter paths in /nix/store/58lissxzppdn2cy62abhfq98ln83lvqd-perl5.28.0-App-ClusterSSH-4.13.2
checking for references to /build in /nix/store/58lissxzppdn2cy62abhfq98ln83lvqd-perl5.28.0-App-ClusterSSH-4.13.2...
shrinking RPATHs of ELF executables and libraries in /nix/store/8308zlb51l4i74xjmdh2k2l4gfqn46i7-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
gzipping man pages under /nix/store/8308zlb51l4i74xjmdh2k2l4gfqn46i7-perl5.28.0-App-ClusterSSH-4.13.2-devdoc/share/man/
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/8308zlb51l4i74xjmdh2k2l4gfqn46i7-perl5.28.0-App-ClusterSSH-4.13.2-devdoc
checking for references to /build in /nix/store/8308zlb51l4i74xjmdh2k2l4gfqn46i7-perl5.28.0-App-ClusterSSH-4.13.2-devdoc...
/nix/store/58lissxzppdn2cy62abhfq98ln83lvqd-perl5.28.0-App-ClusterSSH-4.13.2
```
</p></details>

aanderse commented:
> @aanderse, yes, I'm using it and it works for me. The code you linked is used to find xterm (or some other configured terminal), so the hardcoded paths would be used if it weren't found already in `$PATH`, i.e. it will not find it on NixOS and exit. I guess we could add `/run/current-system/sw/bin` to the hardcoded list, but I think it is sufficient to rely on `$PATH` being set properly and having the terminal in it.

Makes sense to me. In this situation you should avoid adding /run/current-system/sw/bin as it is always better to explicitly declare dependencies when you can, like this situation. Looking good.

Mic92 commented:
@GrahamcOfBorg build perldevelPackages.ClusterSSH

Mic92 commented:
@volth do we need to take special care if packages do not build for specific perl version?

GrahamcOfBorg commented:
<!--REQUEST_ID=88701fc7-d0e8-4ab0-b2c9-49039c58a9a2-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=edb3aa92-0a9b-4e87-85ac-6c49f7fe61cd)

Attempted: perldevelPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/w0gyc5xkj81ywrchsb0gdiazw4r5knn3-perl5.29.4-X11-Protocol-Other-30/lib
patching script interpreter paths in /nix/store/w0gyc5xkj81ywrchsb0gdiazw4r5knn3-perl5.29.4-X11-Protocol-Other-30
checking for references to /build in /nix/store/w0gyc5xkj81ywrchsb0gdiazw4r5knn3-perl5.29.4-X11-Protocol-Other-30...
shrinking RPATHs of ELF executables and libraries in /nix/store/if38vapx8q8nl685gbynhbjin474l9wp-perl5.29.4-X11-Protocol-Other-30-devdoc
gzipping man pages under /nix/store/if38vapx8q8nl685gbynhbjin474l9wp-perl5.29.4-X11-Protocol-Other-30-devdoc/share/man/
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/if38vapx8q8nl685gbynhbjin474l9wp-perl5.29.4-X11-Protocol-Other-30-devdoc
checking for references to /build in /nix/store/if38vapx8q8nl685gbynhbjin474l9wp-perl5.29.4-X11-Protocol-Other-30-devdoc...
cannot build derivation '/nix/store/3q5ccjgvx8l2bksvj5qib54pwjdinbpd-perl5.29.4-App-ClusterSSH-4.13.2.drv': 2 dependencies couldn't be built
error: build of '/nix/store/3q5ccjgvx8l2bksvj5qib54pwjdinbpd-perl5.29.4-App-ClusterSSH-4.13.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=88701fc7-d0e8-4ab0-b2c9-49039c58a9a2-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48857&attempt_id=2cd59fc5-0ba7-4453-82d4-5503b745de97)

Attempted: perldevelPackages.ClusterSSH

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl/5.29.4/aarch64-linux-thread-multi/Tk/pTk/Tcl-pTk: interpreter directive changed from "/usr/bin/perl -w -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl" to "/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/bin/perl -w -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl"
/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl/5.29.4/aarch64-linux-thread-multi/Tk/pTk/mkVFunc: interpreter directive changed from "/usr/bin/perl -w -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl" to "/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/bin/perl -w -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/n22j0jdr8ka9f7x86x0s379mim8afgi5-perl-5.29.4/lib/perl5/site_perl -I/nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034/lib/perl5/site_perl"
checking for references to /build in /nix/store/31ynnqhjxafdvinsvx45q7vxvkllgrqb-perl5.29.4-Tk-804.034...
shrinking RPATHs of ELF executables and libraries in /nix/store/hlzxixmfb9rpr90p8gh391zqhyj8k7bs-perl5.29.4-Tk-804.034-devdoc
gzipping man pages under /nix/store/hlzxixmfb9rpr90p8gh391zqhyj8k7bs-perl5.29.4-Tk-804.034-devdoc/share/man/
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/hlzxixmfb9rpr90p8gh391zqhyj8k7bs-perl5.29.4-Tk-804.034-devdoc
checking for references to /build in /nix/store/hlzxixmfb9rpr90p8gh391zqhyj8k7bs-perl5.29.4-Tk-804.034-devdoc...
cannot build derivation '/nix/store/bm5l0xi19zpg5hkhq8cm9j7h91jvg2rf-perl5.29.4-App-ClusterSSH-4.13.2.drv': 2 dependencies couldn't be built
error: build of '/nix/store/bm5l0xi19zpg5hkhq8cm9j7h91jvg2rf-perl5.29.4-App-ClusterSSH-4.13.2.drv' failed
```
</p></details>

volth commented:
> @volth do we need to take special care if packages do not build for specific perl version?

I'd put ```meta.broken = versionAtLeast perl.version "5.29"```

ryantm commented:
@GrahamcOfBorg build perlPackages.ClusterSSH

ryantm commented:
@GrahamcOfBorg build perldevelPackages.ClusterSSH

aither64 commented:
Sorry for the delay. I've rebased it on nixos-unstable, added `meta.broken` for perl 5.29 and restricted platforms to `x86_64-linux` based on the build failures. Is it suitable for merging? I have no idea what to do about those build failures at this time.

herrwiese commented:
This is obsoleted by  #65043 and should be closed.
